Frameworks/CoreText/CTLine.mm

```diff
@@ -22,25 +22,89 @@
+    CFRange replaceRange = { CFAttributedStringGetLength(attString), 0 };
+    CFAttributedStringReplaceString(attString, replaceRange, CFAttributedStringGetString(appendString));
+
+    CFRange enumerationRange = { 0, CFAttributedStringGetLength(appendString) };
+    CFIndex index = enumerationRange.location;
+    CFIndex prevlen = CFAttributedStringGetLength(attString);
+    CFIndex lenDiff = 0;
+    CFRange currentRange;
+
+    while ((index >= enumerationRange.location) && index < (enumerationRange.location + enumerationRange.length)) {
+        lenDiff = 0;
+
+        CFDictionaryRef attributes = CFAttributedStringGetAttributes(appendString, index, &currentRange);
+
+        currentRange = __CFRangeIntersectionRange(currentRange, enumerationRange);
+
+        CFAttributedStringSetAttributes(attString,
+                                        { currentRange.location + replaceRange.location, currentRange.length },
+                                        attributes,
+                                        true);
+        // account for length changes
+        CFIndex currentlen = CFAttributedStringGetLength(attString);
+        if (prevlen != currentlen) {
+            lenDiff = currentlen - prevlen;
+            prevlen = currentlen;
+        }
+
+        index += (currentRange.length + lenDiff);
+        enumerationRange.length += lenDiff;
+
+        if (currentRange.length == 0) {
+            break;
+        }
+    }
+}
+
+static inline void __CFAtributedStringInsertAttributedStringOneAttribute(CFMutableAttributedStringRef attString,
+                                                                         CFIndex index,
+                                                                         CFStringRef string,
+                                                                         CFDictionaryRef attribs) {
+    CFRange replaceRange = { index, 0 };
+
+    CFAttributedStringReplaceString(attString, replaceRange, string);
+
+    CFRange range = { 0, CFStringGetLength(string) };
+    CFAttributedStringSetAttributes(attString, { range.location + replaceRange.location, range.length }, attribs, true);
+}
+
+static inline void __CFAtributedStringAppendAttributedStringOneAttribute(CFMutableAttributedStringRef attString,
+                                                                         CFStringRef string,
+                                                                         CFDictionaryRef attribs) {
+    __CFAtributedStringInsertAttributedStringOneAttribute(attString, CFAttributedStringGetLength(attString), string, attribs);
+}
+
 static CFStringRef __CTCreateReversedString(CFStringRef string) {
     if (string == nullptr) {
         return nullptr;
     }
 
     CFIndex length = CFStringGetLength(string);
-    CFIndex usedBufLen;
-    CFStringGetBytes(string, CFRangeMake(0, length), kCFStringEncodingUTF16, 0, false, nullptr, length, &usedBufLen);
-
     if (length < 2) {
```

Comment:
LOL

---
In reply to: [112724123](https://github.com/Microsoft/WinObjC/pull/2586#discussion_r112724123) [](ancestors = 112724123)